### ISSUE TITLE
[Question] Drastically different thread_tsan performance on gg grading server & development server

### ISSUE THREAD

**Issue by losif63** [kaist-cp/cs431#835]
I finished implementing homework 3, and I tested it using the scripts/submit.sh in the development server on https://cloud.fearless.systems. In the development server, every time I run the grading script, I always get a 100/100.

However, when I submit the code on the grading website on gg.kaist.ac.kr, I seem to keep getting a timeout error on the optimistic fine-grained thread_tsan test cases. This is extremely frustrating, because I never got a timeout error on the development server. For example, testing optimistic fine-grained stress_concurrent with cargo_tsan on the development server only takes about 6~8 seconds, but for some reason it seems to take more than a minute on the grading server. I tried submitting my code multiple times to check if it is a temporary server issue, but I seem to keep running into the same timeout error.

What is more frustrating is that this timeout error seems to occur irregularly. For example, in my submission history for homework 3, submission ID# 23468 ~ 23470 are the same code. In submission ID#23468, testing optimistic fine-grained stress_concurrent with cargo_tsan took about 9 seconds, which is similar to what I got when running my code on the development server. However, I got a timeout error for the same test in submission ID# 23469~23470, even though I used the same code.

To summarize, I never got a timeout error & almost all tests take less than 10 seconds on the development server, but for some reason I keep getting a timeout error on the grading server in the gg.kaist.ac.kr website. Also, this timeout error seems to happen irregularly. I'm not sure whether this issue is a problem on my side or on the server side, so I decided to post this question here. Is the thread sanitizer used for the development server and the grading server different anyhow? Also, since the main purpose of using sanitizers is to check the correctness of my code (not performance), could you raise the timeout limit for tests using sanitizers?

Thank you.

**Comment by tomtomjhj** [comment]
Try changing Relaxed to SeqCst. Thread sanitizer doesn't understand SeqLock synchronization and will raise false alarms. I'm not sure if this was the actual problem, but you need to fix this anyway.

**Comment by losif63** [comment]
> Try changing Relaxed to SeqCst. Thread sanitizer doesn't understand SeqLock synchronization and will raise false alarms. I'm not sure if this was the actual problem, but you need to fix this anyway.

Thanks for pointing this out! I checked and there actually was an occurrence of Relaxed instead of SeqCst, I must have made a mistake and let it slip through my mind. However, I tried uploading the fixed code onto the gg grading server, but it still doesn't work :( I keep getting a timeout error when testing optimistic_fine_grained::stress_concurrent with the thread sanitizer.

What I can't understand most is the fact that my code never stalls on the development server. As I said before, testing optimistic_fine_grained::stress_concurrent with the thread sanitizer on the development server only takes only 7~9 seconds. If I could replicate what is going on on the grading server during development, I could try to identify what is causing my program to stall, but alas I cannot do so. 

Could there possibly something wrong with the virtual environment the grading server is using?

**Comment by tomtomjhj** [comment]
> It was originally part of the base code, so I didn't touch it until now.

Fixed the skeleton code.

> Could there possibly something wrong with the virtual environment the grading server is using?

It seems that there's some issue with gg. We will investigate.

**Comment by tomtomjhj** [comment]
I couldn't figure out the root cause of the issue, but anyway I increased timeout to 3 mins and reduces number of threads for optimistic_fine_grained tests under tsan. Sorry for the inconvenience.

**Comment by losif63** [comment]
I just checked, and everything seems to be working fine now. Thank you so much for helping me out on this issue!
